File: backend/routers/admin/field_harmonization.py
"""Field harmonization API endpoints for managing field mappings."""

import logging
from typing import Any, Dict, List

from fastapi import APIRouter, HTTPException, Query

from backend.schemas.field_harmonization import (
    CreateMappingRequest,
    FieldCluster,
    FieldDistribution,
    FieldHarmonizationMapping,
    FieldMapping,
    FieldMappingImpact,
    HarmonizationSuggestion,
    SimilarityMatrix,
    SuggestFieldMappingsRequest,
    UpdateMappingRequest,
)
from backend.services.analytics.field_harmonization import (
    apply_active_mappings_retroactively,
    create_field_mapping,
    delete_field_mapping,
    generate_field_similarity_matrix,
    get_active_mappings,
    get_field_distribution,
    get_field_embeddings_umap,
    get_field_mapping_history,
    get_field_mapping_impact,
    invalidate_field_embeddings,
    suggest_field_mappings,
    update_field_mapping,
)

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get("/distribution", response_model=List[FieldDistribution])
async def get_field_distributions(
    min_occurrence: int = 5,
    include_values: bool = True,
    top_values_limit: int = 10,
):
    """Get distribution statistics for fields."""
    return await get_field_distribution(
        min_occurrence=min_occurrence,
        include_values=include_values,
        top_values_limit=top_values_limit,
    )


@router.get("/similarity-matrix", response_model=SimilarityMatrix)
async def get_similarity_matrix(
    min_occurrence: int = 5,
    include_field_values: bool = True,
):
    """Generate a similarity matrix for fields."""
    return await generate_field_similarity_matrix(
        min_occurrence=min_occurrence,
        include_field_values=include_field_values,
    )


@router.post("/suggest-mappings", response_model=HarmonizationSuggestion)
async def suggest_mappings(
    request: SuggestFieldMappingsRequest,
):
    """Suggest field mappings based on similarity."""
    return await suggest_field_mappings(
        similarity_threshold=request.similarity_threshold,
        min_occurrence=request.min_occurrence,
        protected_fields=request.protected_fields,
    )


@router.get("/mappings", response_model=List[FieldHarmonizationMapping])
async def get_mappings(
    days: int = 30,
):
    """Get field mapping history."""
    return await get_field_mapping_history(days=days)


@router.get("/mappings/active", response_model=List[FieldHarmonizationMapping])
async def get_active_mapping():
    """Get all active field mappings."""
    return await get_active_mappings()


@router.post("/mappings/impact", response_model=FieldMappingImpact)
async def calculate_mapping_impact(
    field_mapping: Dict[str, str],
):
    """Calculate the impact of applying a field mapping."""
    if not field_mapping:
        raise HTTPException(status_code=400, detail="Field mapping cannot be empty")

    return await get_field_mapping_impact(field_mapping)


@router.get("/embeddings/umap", response_model=Dict[str, Any])
async def get_embeddings_umap(
    min_occurrence: int = Query(5, description="Minimum occurrences for a field to be included"),
    n_neighbors: int = Query(15, description="UMAP: Number of neighbors"),
    min_dist: float = Query(0.1, description="UMAP: Minimum distance"),
    n_components: int = Query(3, description="UMAP: Number of dimensions (2 or 3)", ge=2, le=3),
    metric: str = Query("cosine", description="UMAP: Distance metric"),
):
    """Generate UMAP projection of field embeddings."""
    result = await get_field_embeddings_umap(
        min_occurrence=min_occurrence,
        n_neighbors=n_neighbors,
        min_dist=min_dist,
        n_components=n_components,
        metric=metric,
    )
    if not result or not result.get("fields"):
        raise HTTPException(status_code=500, detail="Failed to generate UMAP projection")
    return result


@router.post("/embeddings/invalidate", status_code=200, response_model=Dict[str, Any])
async def invalidate_embeddings():
    """Invalidate and delete all stored field embeddings."""
    deleted_count = await invalidate_field_embeddings()
    if deleted_count < 0:
        raise HTTPException(status_code=500, detail="Failed to invalidate embeddings")
    return {"message": f"Successfully invalidated {deleted_count} embeddings."}


@router.post("/mappings", response_model=FieldHarmonizationMapping)
async def create_mapping(
    request: CreateMappingRequest,
):
    """Create a new field mapping."""
    mapping = await create_field_mapping(
        name=request.name,
        mappings=request.mappings,
        description=request.description,
        is_active=request.is_active,
        created_by=request.created_by,
    )

    if not mapping:
        raise HTTPException(status_code=500, detail="Failed to create field mapping")

    return mapping


@router.patch("/mappings/{mapping_id}", response_model=FieldHarmonizationMapping)
async def update_mapping(
    mapping_id: str,
    request: UpdateMappingRequest,
):
    """Update an existing field mapping."""
    mapping = await update_field_mapping(
        mapping_id=mapping_id,
        name=request.name,
        description=request.description,
        mappings_to_add=request.mappings_to_add,
        mappings_to_remove=request.mappings_to_remove,
        is_active=request.is_active,
    )

    if not mapping:
        raise HTTPException(status_code=404, detail=f"Mapping {mapping_id} not found")

    return mapping


@router.delete("/mappings/{mapping_id}", status_code=204)
async def delete_mapping_endpoint(
    mapping_id: str,
):
    """Delete a field harmonization mapping permanently."""
    success, message = await delete_field_mapping(mapping_id)

    if not success:
        # Determine the appropriate status code based on the message
        if "not found" in message.lower():
            raise HTTPException(status_code=404, detail=message)
        else:
            # Assume internal server error for other failures
            raise HTTPException(status_code=500, detail=message)

    # On success, return None and FastAPI will send a 204 response
    return None


@router.post("/mappings/apply-retroactive", response_model=Dict[str, Any])
async def apply_mappings_retroactively_endpoint(
    batch_size: int = Query(500, ge=100, le=5000, description="Number of listings to process per batch"),
):
    """Apply all active mappings to existing analyzed listings."""
    try:
        result = await apply_active_mappings_retroactively(batch_size=batch_size)
        if "error" in result:
            # Distinguish between client errors (e.g., conflicts) and server errors
            # For now, assume internal error
            raise HTTPException(status_code=500, detail=result["error"])
        return result
    except Exception as e:
        logger.error(f"Unhandled error during retroactive mapping application: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail="An unexpected error occurred.")

File: frontend/src/lib/api/admin/field-harmonization.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import apiClient from '@/lib/api-client';
import { useToast } from '@/hooks/use-toast';

// Import query keys for invalidation
import { QUERY_API_QUERY_KEY_PREFIX } from '@/lib/api/query/query';
import { PRICE_STATS_QUERY_KEY_PREFIX } from '@/lib/api/analytics/price-stats';

// --- Define and Export local types
export type FieldType =
  | "numeric"
  | "categorical"
  | "boolean"
  | "text"
  | "mixed"
  | "unknown";

export interface FieldValuePattern {
  type: FieldType;
  unit?: string;
  distinct_values?: number;
  value_examples?: string[];
  stats?: Record<string, any>;
}

export interface FieldDistribution {
  field_name: string;
  occurrence_count: number;
  distinct_value_count: number;
  top_values: [string, number][];
  field_type: FieldType;
  pattern: FieldValuePattern | null;
  cluster_id?: string;
}

export interface FieldMapping {
    original_field: string;
    target_field: string;
}

export interface FieldHarmonizationMapping {
  _id: string;
  name: string;
  description?: string;
  mappings: FieldMapping[];
  created_at: string;
  updated_at: string;
  is_active: boolean;
  created_by?: string;
  affected_listings: string[];
}

export interface SimilarityMatrix {
  fields: string[];
  scores: number[][];
  timestamp: string;
}

export interface SuggestMappingsPayload {
  similarity_threshold?: number;
  min_occurrence?: number;
  protected_fields?: string[];
}

export interface HarmonizationSuggestion {
    clusters: any[]; // Replace with actual FieldCluster type if needed
    suggested_mapping: Record<string, string>;
    impact: any; // Replace with actual FieldMappingImpact type if needed
    similarity_threshold: number;
    created_at: string;
}

export interface UmapFieldData {
  name: string;
  type: FieldType;
  count: number;
  coordinates: [number, number] | [number, number, number]; // Supports 2D or 3D
}

export interface FieldEmbeddingsUmapResponse {
  fields: UmapFieldData[];
  umap_params: {
    n_neighbors: number;
    min_dist: number;
    n_components: number;
    metric: string;
  };
}

// --- Define Request Payload Types ---

// Matches backend/schemas/field_harmonization.py::CreateMappingRequest
export interface CreateMappingPayload {
  name: string;
  description?: string;
  mappings: FieldMapping[];
  is_active?: boolean;
  created_by?: string;
}

// Matches backend/schemas/field_harmonization.py::UpdateMappingRequest
export interface UpdateMappingPayload {
  name?: string;
  description?: string;
  mappings_to_add?: FieldMapping[];
  mappings_to_remove?: string[]; // List of original_field names
  is_active?: boolean;
}

// Matches backend/routers/analytics/field_harmonization.py::calculate_mapping_impact payload
export interface CalculateImpactPayload {
  field_mapping: Record<string, string>;
}

// Matches backend/routers/analytics/field_harmonization.py::calculate_mapping_impact response
export interface FieldMappingImpactResponse {
    // Define based on backend schema backend/schemas/field_harmonization.py::FieldMappingImpact
    total_fields: number;
    total_mapped_fields: number;
    total_affected_documents: number;
    field_impacts: Record<string, Record<string, any>>;
    clusters: any[]; // Replace with FieldCluster if needed
    before_after_samples: Record<string, any>[];
}

// --- API Client Methods ---

const basePath = '/admin/field-harmonization'; // Update base path
const FIELD_HARMONIZATION_QUERY_KEY_PREFIX = 'fieldHarmonization';

// --- Query Hooks ---

export function useFieldDistribution(
    minOccurrence: number = 5,
    includeValues: boolean = true,
    topValuesLimit: number = 10
) {
  return useQuery<FieldDistribution[], Error>({
    queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'distribution', { minOccurrence, includeValues, topValuesLimit }],
    queryFn: () => {
      const params = new URLSearchParams();
      params.append('min_occurrence', minOccurrence.toString());
      params.append('include_values', includeValues.toString());
      params.append('top_values_limit', topValuesLimit.toString());
      return apiClient._fetch(`${basePath}/distribution?${params}`);
    },
  });
}

export function useFieldMappings(days: number = 30) {
  return useQuery<FieldHarmonizationMapping[], Error>({
    queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'mappings', { days }],
    queryFn: () => {
      const params = new URLSearchParams();
      params.append('days', days.toString());
      return apiClient._fetch(`${basePath}/mappings?${params}`);
    }
  });
}

export function useActiveFieldMapping() {
  return useQuery<FieldHarmonizationMapping[], Error>({ // Backend returns a list
    queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'mappings', 'active'],
    queryFn: async () => {
        try {
            return await apiClient._fetch(`${basePath}/mappings/active`);
        } catch (error: any) {
            // Allow other errors to propagate
            console.error("Error fetching active mapping:", error);
            throw error;
        }
    }
  });
}

export function useSimilarityMatrix(
    minOccurrence: number = 5,
    includeFieldValues: boolean = true // Renamed to match backend
) {
  return useQuery<SimilarityMatrix, Error>({
    queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'similarityMatrix', { minOccurrence, includeFieldValues }],
    queryFn: () => {
      const params = new URLSearchParams();
      params.append('min_occurrence', minOccurrence.toString());
      params.append('include_field_values', includeFieldValues.toString());
      return apiClient._fetch(`${basePath}/similarity-matrix?${params}`);
    }
  });
}

export function useFieldEmbeddingsUmap(
  minOccurrence: number = 5,
  nNeighbors: number = 15,
  minDist: number = 0.1,
  nComponents: number = 2,
  metric: string = 'cosine'
) {
  return useQuery<FieldEmbeddingsUmapResponse, Error>({
    queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'embeddings', 'umap', { minOccurrence, nNeighbors, minDist, nComponents, metric }],
    queryFn: () => {
      const params = new URLSearchParams();
      params.append('min_occurrence', minOccurrence.toString());
      params.append('n_neighbors', nNeighbors.toString());
      params.append('min_dist', minDist.toString());
      params.append('n_components', nComponents.toString());
      params.append('metric', metric);
      return apiClient._fetch(`${basePath}/embeddings/umap?${params}`);
    }
  });
}

// --- Mutation Hooks ---

export function useCreateFieldMappingMutation() {
  const queryClient = useQueryClient();
  const { toast } = useToast();
  return useMutation<FieldHarmonizationMapping, Error, CreateMappingPayload>({
    mutationFn: (mapping) => apiClient._fetch(`${basePath}/mappings`, {
        method: 'POST',
        body: JSON.stringify(mapping),
    }),
    onSuccess: (data) => {
      queryClient.invalidateQueries({ queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'mappings'] });
      // queryClient.invalidateQueries({ queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'mappings', 'active'] }); // Covered by broader invalidation
      toast({ title: "Success", description: `Mapping "${data.name}" created.` });
    },
    onError: (error) => {
      toast({ variant: "destructive", title: "Error", description: error.message || "Failed to create mapping." });
    },
  });
}

export function useUpdateFieldMappingMutation() {
  const queryClient = useQueryClient();
  const { toast } = useToast();
  return useMutation<FieldHarmonizationMapping, Error, { mappingId: string; updates: UpdateMappingPayload }>({
    mutationFn: ({ mappingId, updates }) => apiClient._fetch(`${basePath}/mappings/${mappingId}`, {
        method: 'PATCH',
        body: JSON.stringify(updates),
    }),
    onSuccess: (data) => {
      queryClient.invalidateQueries({ queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'mappings'] });
      toast({ title: "Success", description: `Mapping "${data.name}" updated.` });
    },
    onError: (error) => {
      toast({ variant: "destructive", title: "Error", description: error.message || "Failed to update mapping." });
    },
  });
}

export function useDeleteFieldMappingMutation() {
  const queryClient = useQueryClient();
  const { toast } = useToast();
  return useMutation<void, Error, string>({
    mutationFn: (mappingId) => apiClient._fetch(`${basePath}/mappings/${mappingId}`, {
        method: 'DELETE',
    }), // No body needed for DELETE, returns 204
    onSuccess: (_, mappingId) => {
      queryClient.invalidateQueries({ queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'mappings'] });
      toast({ title: "Success", description: `Mapping deleted.` }); // Name might not be available here easily
    },
    onError: (error) => {
      toast({ variant: "destructive", title: "Error", description: error.message || "Failed to delete mapping." });
    },
  });
}

export function useFieldMappingImpactMutation() {
  const { toast } = useToast();
  return useMutation<FieldMappingImpactResponse, Error, CalculateImpactPayload>({
    mutationFn: (payload) => apiClient._fetch(`${basePath}/mappings/impact`, {
        method: 'POST',
        body: JSON.stringify(payload.field_mapping), // Backend expects the dict directly
    }),
    onSuccess: () => {
      toast({ title: "Success", description: "Field mapping impact calculated." });
    },
    onError: (error) => {
      toast({ variant: "destructive", title: "Error", description: error.message || "Failed to calculate impact." });
    },
  });
}

export function useSuggestFieldMappingsMutation() {
  const { toast } = useToast();
  return useMutation<HarmonizationSuggestion, Error, SuggestMappingsPayload | undefined>({ // Payload can be empty for defaults
    mutationFn: (payload = {}) => {
        const body = {
            similarity_threshold: payload.similarity_threshold ?? 0.75,
            min_occurrence: payload.min_occurrence ?? 5,
            protected_fields: payload.protected_fields ?? [],
        };
        return apiClient._fetch(`${basePath}/suggest-mappings`, {
            method: 'POST',
            body: JSON.stringify(body),
        });
    },
    onSuccess: () => {
      toast({ title: "Success", description: "Field mapping suggestions generated." });
    },
    onError: (error) => {
      toast({ variant: "destructive", title: "Error", description: error.message || "Failed to suggest mappings." });
    },
  });
}

export function useInvalidateFieldEmbeddingsMutation() {
  const queryClient = useQueryClient();
  const { toast } = useToast();
  return useMutation<{ message: string }, Error, void>({
    mutationFn: () => apiClient._fetch(`${basePath}/embeddings/invalidate`, {
        method: 'POST',
    }),
    onSuccess: (data) => {
      queryClient.invalidateQueries({ queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'embeddings', 'umap'] });
      toast({ title: "Success", description: data.message || "Field embeddings invalidated." });
    },
    onError: (error) => {
      toast({ variant: "destructive", title: "Error", description: error.message || "Failed to invalidate embeddings." });
    },
  });
}

export function useApplyMappingsRetroactivelyMutation() {
  const queryClient = useQueryClient();
  const { toast } = useToast();
  return useMutation<{ message: string, processed_count: number, total_updated: number }, Error, number | undefined> ({
    mutationFn: (batchSize) => { // batchSize can be undefined for default
        const params = new URLSearchParams();
        if (batchSize !== undefined) {
            params.append('batch_size', batchSize.toString());
        }
        return apiClient._fetch(`${basePath}/mappings/apply-retroactive?${params.toString()}`, {
            method: 'POST',
            // No body needed for this trigger
        });
    },
    onSuccess: (data) => {
      queryClient.invalidateQueries({ queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'mappings'] });
      queryClient.invalidateQueries({ queryKey: [FIELD_HARMONIZATION_QUERY_KEY_PREFIX, 'distribution'] });
      queryClient.invalidateQueries({ queryKey: [QUERY_API_QUERY_KEY_PREFIX, 'listings'] }); 
      queryClient.invalidateQueries({ queryKey: [PRICE_STATS_QUERY_KEY_PREFIX, 'priceStats'] }); 
      toast({ title: "Success", description: data.message || "Retroactive mapping application started." });
    },
    onError: (error) => {
      toast({ variant: "destructive", title: "Error", description: error.message || "Failed to apply mappings retroactively." });
    },
  });
}